### ISSUE TITLE
TensorFlow binary crashes on Apple M1 in x86_64 Docker container

### ISSUE THREAD

**Issue by dwyatte** [tensorflow/tensorflow#52845]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): TensorFlow 2.6.0, tf-nightly 2.8.0.dev20211028
- Python version: 3.6.9, 3.7.x, 3.8.x
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
```
dwyatte-macbookpro:~ dwyatte$ docker run tensorflow/tensorflow:latest python -c "import tensorflow as tf"    
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
2021-10-28 22:50:41.481158: F tensorflow/core/lib/monitoring/sampler.cc:42] Check failed: bucket_limits_[i] > bucket_limits_[i - 1] (0 vs. 10)
qemu: uncaught target signal 6 (Aborted) - core dumped
```

**Describe the expected behavior**
Clean exit

**Standalone code to reproduce the issue**
Requires an Apple M1 (arm64) host OS:
`docker run tensorflow/tensorflow:latest python -c "import tensorflow as tf"`

This was previously mentioned in https://github.com/tensorflow/tensorflow/issues/42387 but unfortunately closed. When importing TensorFlow in an x86_64 docker container on an Apple M1, TensorFlow crashes. As far as I can tell, this should work as I can import and use other Python packages in the same container without problems (including things like `numpy`).

It's unclear whether this is something that can be avoided at the TensorFlow level or an unavoidable bug in qemu ([[1]](https://github.com/docker/for-mac/issues/5342#issuecomment-779133157), [[2]](https://gitlab.com/qemu-project/qemu/-/issues/601)), but I wanted to reraise the issue.

**Comment by mohantym** [comment]
Hi @sanatmpa1! Could you look at this issue?

**Comment by vazkir** [comment]
I am taking a class where we use tensorflow inside docker containers and everybody with an M1 mac in that class had this exact same issue including me. Unfortunately nobody has found a fixso I am going to subsribe to this issue as well, I hope there exist some kind of workarround/solution!

**Comment by alexcombessie** [comment]
Hi,

I have the exact same issue. It is hindering my development process. While my app is deployed on an x86 server, I do need to use my M1 mac with emulation to develop code locally and to push it to production.

All other major data science packages work correctly under x86 rosetta emulation: pandas, scikit-learn, torch, transformers, spacy, xgboost, lightgbm.

I appreciate the great work you are doing with TensorFlow. I would be really grateful if you could take the time to help the data scientists / ML engineers out there who are using ARM-based development laptops.

Thanks a lot,

Alex

PS: I am not interested in forks like tensorflow-macos etc as I need my work to be cross-platform.

**Comment by bhack** [comment]
https://github.com/apple/tensorflow_macos/issues/164#issuecomment-776785984

https://github.com/ARM-software/Tool-Solutions/tree/master/docker/tensorflow-aarch64

But as someone still needs to use this in emulation I suppose in that It could be a qemu BUG with `DBL_MAX` in emulation

**Comment by vazkir** [comment]
Did anybody find anyway to run tensorflow inside a docker container on any M1, M1 Pro or M1 Max device? Would really love to know any workaround so I can start building containers with tf. Thanks in advance for any tips!

**Comment by bhack** [comment]
If the point is to have a published X86 wheel without AVX we have already an open ticket, so it is better to add a comment there instead of having a new ticket:

https://github.com/tensorflow/tensorflow/issues/19584

If instead you want to have AVX TCG support in QEMU e.g. on M1 there is already an open ticket at:
https://gitlab.com/qemu-project/qemu/-/issues/164

**Comment by dwyatte** [comment]
So I do think this is due to AVX instructions. If I install an unofficial wheel (e.g., from https://github.com/yaroslavvb/tensorflow-community-wheels/issues/198) and run a variant of the `docker run` command above, I do not get a crash on import.

```
dwyatte-macbookpro:~ dwyatte$ docker run -it tensorflow/tensorflow:latest bash -c 'pip uninstall -y tensorflow-cpu && pip install -U https://tf.novaal.de/barcelona/tensorflow-2.6.0-cp38-cp38-linux_x86_64.whl && python -c "import tensorflow as tf; tf.print(\"hello world\")"'
...
2021-11-15 23:44:35.660302: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
hello world
```

Thanks for the lead @bhack. I agree, some solutions which you mention are:
1.) Publishing non-AVX wheels (or having non-AVX code paths available within a single wheel)
2.) Correctly handling in qemu via emulation/TCG/etc.

**Comment by bhack** [comment]
For the first point I don't know if anyone at @intel-tensorflow is interested to publish an SSE4.x only wheel in https://pypi.org/project/intel-tensorflow/

**Comment by gabac** [comment]
@dwyatte Thanks a lot for the tip. With an unofficial wheel I was able to get Tensorflow running within Docker on an Apple M1 processor 🚀

**Comment by janvdp** [comment]
@gabac One you built or one that is available online? I'm facing the same issue...

**Comment by gabac** [comment]
E.g. if you use pip as a package manager use e.g. `pip install -U https://tf.novaal.de/barcelona/tensorflow-2.5.0-cp37-cp37m-linux_x86_64.whl` for Python 3.7, Tensorflow 2.5.0

**Comment by janvdp** [comment]
Thanks, that did the trick! Unfortunately, Docker + M1 Mac seems to be pretty slow... :( (not talking about training...)

**Comment by bhack** [comment]
For the performance you need to use `tensorflow-macos` in a arm64v8 image e.g. https://hub.docker.com/r/arm64v8/ubuntu/

**Comment by bhack** [comment]
Untill we will have `AMX2` support in the compiler stack (starting with LLVM) It will be hard to have a docker Linux image with a TF optimized for M1 CPU.

See 
> Special note about Apple M1 https://nod.ai/nod-codegen-beats-mkl-blis-accelerate/

**Comment by janvdp** [comment]
It seems to be Docker Desktop on the M1 that is causing performance issues. It's not even TF related, only testing model conversion and single inference...

**Comment by bhack** [comment]
Docker desktop on Apple silicon for X86 i think uses rosetta2.

https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment

> Rosetta translates all x86_64 instructions, but it doesn’t support the execution of some newer instruction sets and processor features, such as AVX, AVX2, and AVX512 vector instructions. If you include these newer instructions in your code, execute them only after verifying that they are available. For example, to determine if AVX512 vector instructions are available, use the sysctlbyname function to check the hw.optional.avx512f attribute.


Other then the missing CPU accell you will not have Neural engine and GPU access.

So I don't think it could be suggested to use It in emulation on linux for performance as the platform isn't documented, fully supported on linux and in the opensoruce compilers stacks.

**Comment by kennysong** [comment]
(Thanks for the tip about the non-AVX unofficial build, will give that a try. 1 or 2 mentioned by @dwyatte seem like better medium term solutions though.)

I think Docker uses QEMU for emulation (not rosetta2), which is one of the reasons it's slow in general. 

[1] https://docs.docker.com/desktop/mac/apple-silicon/#known-issues
[2] https://news.ycombinator.com/item?id=25449561

**Comment by bhack** [comment]
Yes It Is just that Partially Docker itself still run Rosetta 2 as some binaries are still Darwin/AMD but the emulation is Qemu.

But both are no AVX support so I hard to achieve top performances and Linux don't support AMX2 as are still going to be supported at compiler level (and fully documented/reverse engineered).

**Comment by janvdp** [comment]
I'm building the software on my Mac (old one for now), where I mostly build the apps, APIs, UI, inference engines, and run tests for the whole system. I need a Mac to run the model conversion tests for CoreML. 

I tried running linux/amd64 images on the M1  and got things to work for PyTorch and TF (using the non-AVX TF build described above). 

Unfortunately, the tests now take about 3x as long to run, note these are integration tests and don't involve training networks, for which I use a Linux machine. I just wanted to keep them separate in order to run the full suite of tests on linux (excluding coreML tests) while continue to develop on my Mac.

**Comment by DrChrisLevy** [comment]
@janvdp have you made any more progress in the last 2 months ? I added the --platform argument to my docker file to run Linux/amd and installed an unofficial non avx tensor flow 2.7. Finally got the image built but damn it's slow. The unit tests are so much slower. Also many of the tests fail because for some reason some of the rounded precision assertions are different etc. 


If you have made any progress let me know !

**Comment by coreation** [comment]
@DrChrisLevy I'm a novice when it comes to Docker, how did you add the dependency of the unofficial tensor flow to your docker build? Currently I have TF configured as below

```
dependencies:
  - sqlalchemy==1.3.20
  - pymysql==1.0.2
  - pip==21.3
  - pip:
    - tensorflow==2.3.1
    - tensorflow-serving-api==2.3.0
    - sentry-sdk==1.4.3
    - transformers==4.11.3
    - torch==1.9.1
```

**Comment by DrChrisLevy** [comment]
@coreation

So update since I last wrote. Last time above I said I got it working using emulation (so running on x86). This is not ideal because it's slow. It's much faster if you can get it running on aarch64.  It all depends on what other libraries 
you are installing and what your base docker image looks like. 

What docker image are you using?
When you type `uname -m` within the docker container running do you get `aarch64` or `x86_64`? 

If you are doing something like `pip install tensorflow==2.3.1` its going to use one of the official tensorflow wheels. 
And which one it uses depends on if the docker container is running `aarch64` or `x86_64`. But both will have issues. 

If you are running `x86_64` then its gonna be slow. But to get it working you need to install unofficial community wheel without AVX. You can see a list of such options [here](https://github.com/yaroslavvb/tensorflow-community-wheels/issues). For example, when I was trying to go this `x86_64` emulation route I first tried
`TensorFlow 2.7.0 No AVX, No GPU, Python 3.7, 3.8, 3.9, Ubuntu 18.04, multiple Archs` which is this [link](https://github.com/yaroslavvb/tensorflow-community-wheels/issues/206). I tried the `barcelona` one and all of its builds are [here](https://tf.novaal.de/barcelona/). You then have to choose the one for your TF version and Python version. I tried TF  2.7 and Python 3.8 so I used `tensorflow-2.7.0-cp38-cp38-linux_x86_64.whl` which is this link: `https://tf.novaal.de/barcelona/tensorflow-2.7.0-cp38-cp38-linux_x86_64.whl`. So to install this with pip you do
`pip install -f https://tf.novaal.de/barcelona/tensorflow-2.7.0-cp38-cp38-linux_x86_64.whl`. 

I don't recommend this route ^^ though because emulation on m1 chip not ideal. Better to get your docker container running
on `aarch64`. For example I use the docker image `FROM python:3.8-slim`. And I am using an aarch64 community wheel
for Tensorflow. For example, in dockerfile `RUN pip install tensorflow -f https://tf.kmtea.eu/whl/stable.html # Unofficial Tensorflow AARCH64`. I got it from [here](https://github.com/KumaTea/tensorflow-aarch64).

Now in our DEV/PROD ENV we actually manage all dependencies with [poetry](https://python-poetry.org/) and are running everything on x86 so we actually just use the official tensorflow wheels. Our apps dont run on aarch64 in production. But since I use a macbook with m1 chip for local development I needed to get my local dev env running on m1 chip. And its great/fast if I use `aarch64`. So I had to uninstall tensorflow official from my container and then use the above mentioned hack of installing Tensorflow AARCH64. But its working great. 

Let me know if you need more details.

**Comment by coreation** [comment]
Hi @DrChrisLevy Thank you so much for your elaborate reply, I'll try as good as I can to provide answers so that I can hopefully get the docker containers running again.

I believe the base image starts from "continuumio/miniconda3" and then adds dependencies to build the images that we're using. The result of the uname -m is ```x86_64 ```.

I then built our images using the aarch64 tensorflow wheel, I added tensorflow-serving-api as I think it's more basic python implementation to talk with tensorflow-serving.

My build currently doesn't start and returns a more generic error message: ```Error response from daemon: configured logging driver does not support reading``` So I'll have to debug that first.

Do you happen to know where I could find / start looking for a tensorflow-serving M1 compatible image?

**Comment by dwyatte** [comment]
Echoing some of the sentiment here. The slowdown in non-AVX TensorFlow wheels is likely mainly due to emulation (compared to anything inherent to TensorFlow itself) so it would be great to have official aarch64 wheels on PyPI that can use platform detection to `pip install tensorflow` without having to specify a community webpage. This would also support the workflows being discussed here (running aarch64 locally and building/deploying an amd64 image in production)

This issue might be better for discussing aarch64 wheels: https://github.com/tensorflow/tensorflow/issues/52973

Here are some benchmarks for a few of my unit tests (roughly an order of magnitude slower under emulation):

```
dwyatte-macbookpro:tensorflow-test dwyatte$ ./build_and_test.sh 
[+] Building 0.9s (8/8) FINISHED                                                                                                                                                                                        
 => [internal] load build definition from Dockerfile_amd64                                                                                                                                                         0.0s
 => => transferring dockerfile: 209B                                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                                                                                    0.0s
 => [internal] load metadata for docker.io/library/python:3.7-slim                                                                                                                                                 0.8s
 => [1/4] FROM docker.io/library/python:3.7-slim@sha256:71287598b4d9fcc01fa3949035aeace14c6bde733462bb4f64fb4ee13c6b3fec                                                                                           0.0s
 => CACHED [2/4] RUN pip install https://tf.novaal.de/westmere/tensorflow-2.7.0-cp37-cp37m-linux_x86_64.whl                                                                                                        0.0s
 => CACHED [3/4] RUN pip install pandas                                                                                                                                                                            0.0s
 => CACHED [4/4] RUN pip install pytest                                                                                                                                                                            0.0s
 => exporting to image                                                                                                                                                                                             0.0s
 => => exporting layers                                                                                                                                                                                            0.0s
 => => writing image sha256:af60a6f5973c2a8568619c233580d5046670c3dfa5633e287e20869a99a9365e                                                                                                                       0.0s
 => => naming to docker.io/library/tensorflow-test                                                                                                                                                                 0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
================================================================================================= test session starts ==================================================================================================
platform linux -- Python 3.7.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /host
collected 19 items                                                                                                                                                                                                     

tests/test_callbacks.py ...........                                                                                                                                                                              [ 57%]
tests/test_datasets.py ...                                                                                                                                                                                       [ 73%]
tests/test_layers.py .....                                                                                                                                                                                       [100%]

=================================================================================================== warnings summary ===================================================================================================
../usr/local/lib/python3.7/site-packages/flatbuffers/compat.py:19
  /usr/local/lib/python3.7/site-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================ 19 passed, 1 warning in 3.17s =============================================================================================
[+] Building 0.5s (8/8) FINISHED                                                                                                                                                                                        
 => [internal] load build definition from Dockerfile_aarch64                                                                                                                                                       0.0s
 => => transferring dockerfile: 474B                                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                                  0.0s
 => => transferring context: 2B                                                                                                                                                                                    0.0s
 => [internal] load metadata for docker.io/library/python:3.7-slim                                                                                                                                                 0.4s
 => [1/4] FROM docker.io/library/python:3.7-slim@sha256:71287598b4d9fcc01fa3949035aeace14c6bde733462bb4f64fb4ee13c6b3fec                                                                                           0.0s
 => CACHED [2/4] RUN pip install https://snapshots.linaro.org/ldcg/python/tensorflow-io-manylinux/7/tensorflow-io-gcs-filesystem/tensorflow_io_gcs_filesystem-0.21.0-cp37-cp37m-manylinux_2_17_aarch64.manylinux2  0.0s
 => CACHED [3/4] RUN pip install pandas                                                                                                                                                                            0.0s
 => CACHED [4/4] RUN pip install pytest                                                                                                                                                                            0.0s
 => exporting to image                                                                                                                                                                                             0.0s
 => => exporting layers                                                                                                                                                                                            0.0s
 => => writing image sha256:f1a848557f756bf7c99fdbe7bf666d3cb4272eec73a63bdd8abe20283e492aef                                                                                                                       0.0s
 => => naming to docker.io/library/tensorflow-test                                                                                                                                                                 0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
================================================================================================= test session starts ==================================================================================================
platform linux -- Python 3.7.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /host
collected 19 items                                                                                                                                                                                                     

tests/test_callbacks.py ...........                                                                                                                                                                              [ 57%]
tests/test_datasets.py ...                                                                                                                                                                                       [ 73%]
tests/test_layers.py .....                                                                                                                                                                                       [100%]

=================================================================================================== warnings summary ===================================================================================================
../usr/local/lib/python3.7/site-packages/flatbuffers/compat.py:19
  /usr/local/lib/python3.7/site-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================ 19 passed, 1 warning in 0.32s =============================================================================================
```

**Comment by ibash** [comment]
Thanks @DrChrisLevy!

It also turns out tensorflow has [Dockerfiles for arm64v8](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/dockerfiles/dockerfiles/arm64v8), but they don't push the images, so you need to build them yourself. Useful for trying to keep development docker images as close as possible to the official ones.

The one gotcha is to remove the `enum34` python package from the dockerfile.

For posterity:

```
git clone https://github.com/tensorflow/tensorflow.git --depth 1
cd tensorflow/tensorflow/tools/dockerfiles

# edit arm64v8/devel-cpu-arm64v8-jupyter.Dockerfile and remove enum34

docker build -f ./arm64v8/devel-cpu-arm64v8-jupyter.Dockerfile -t tf-devel-cpu-arm64v8-jupyter .
```

Then make a new Dockerfile, and use this:
```
FROM tf-devel-cpu-arm64v8-jupyter

RUN pip install tensorflow -f https://tf.kmtea.eu/whl/stable.html
```

**Comment by coreation** [comment]
@ibash thank you for the heads up, when I ran the docker build using the arm64v8 (not jupyter) I ran into the issue of pandas not being able to install because of the numpy version that was locked in the build file. Setting the version of pandas to something arbitrary like 1.0.5 did the trick.

Do you happen to know if a similar approach exists for tensorflow-serving? That's also a package that currently doesn't run on Apple Silicon.

**Comment by ibash** [comment]
@coreation not sure about tensorflow-serving, I'm not using it.

Re: numpy I hit that too, updated the comment above.

**Comment by coreation** [comment]
@ibash thanks for the comment anyway, from what I read there's no Apple Silicon compatible wheel from the community, but I wanted to check anyway. Thanks for the edit!

**Comment by dakl** [comment]
Are there any update here? Not being able to run tensorflow in docker on our new powerful M1-based machines is both a big downer and quite the annoyance to "solve". ATM our best workaround is having an intel-based VM for development, which feels quite unnecessary.

**Comment by hoangmt** [comment]
This works.
> FROM tf-devel-cpu-arm64v8-jupyter
> 
> # Or use this version: RUN pip install tensorflow-aarch64 -f https://tf.kmtea.eu/whl/stable.html
> 
> RUN pip install tensorflow -f https://tf.kmtea.eu/whl/stable.html
> ```
However, tensorflow does not find gpus.

**Comment by dakl** [comment]
@hoangmt So what do I put in my Pipfile, if I want a version that I can install both in docker on my M1 machine, and build on linux/x64 to send to for example Vertex AI as a training job?

Locally, I don't care about GCP access, it just needs to work so I can develop models with a tiny version of my dataset. The M1-ish GPUs are not good for _training_ models anyway. The "Neural Engine" only accelerates inference AFAIK.

**Comment by JakeFDev** [comment]
If you are using an old 1.x version of Tensorflow, downgrading to max 1.5, Python 3.6, and running docker targeting x86 is a quick, crappy solution since AVX was introduced in 1.6. 

We don't really use Tensorflow for anything critical (was kinda a devs pet project a few years back) so it works for us. Sounds like either qemu needs to support avx to move this foward OR Tensorflow starts publishing multi arch wheels without avx?

**Comment by brunopistone** [comment]
Hello. right now it is still an issue and seems impossible to run official tensorflow images from DockerHub by using Apple M1 chips.
All the solutions proposed are workarounds (use unofficial tensorflow modules). On the basis of the tensorflow version choosen, I have different errors:

Tensorflow 2.5:
```
6vazjvn2w2-algo-1-amynk  | 2022-08-03 16:17:08.519060: F tensorflow/core/lib/monitoring/sampler.cc:42] Check failed: bucket_limits_[i] > bucket_limits_[i - 1] (0 vs. 10)
6vazjvn2w2-algo-1-amynk  | qemu: uncaught target signal 6 (Aborted) - core dumped
```

Tensorflow 2.9:
```
nggi0qa6ib-algo-1-x8ws7  | The TensorFlow library was compiled to use AVX instructions, but these aren't available on your machine.
nggi0qa6ib-algo-1-x8ws7  | qemu: uncaught target signal 6 (Aborted) - core dumped
```

The ticket was opened 9 months ago, can you please share if a solution in order to execute official tensorflow docker images is in roadmap? Right now most of the ML frameworks can be used on Apple M1 chips.

Thank you

**Comment by hanaa-rft** [comment]
I've tried `Tensorflow 2.3.1 ` and I still get 
`F tensorflow/core/lib/monitoring/sampler.cc:42] Check failed: bucket_limits_[i] > bucket_limits_[i - 1] (0 vs. 10)
qemu: uncaught target signal 6 (Aborted) - core dumped`
Any suggestions would be great - thanks.

**Comment by stephenonethree** [comment]
Looks like this problem also comes up if you want to run the GCP Deep Learning Container gcr.io/deeplearning-platform-release/tf-gpu.2-9 on an M1 Mac since it has Tensorflow. The error message is:

```
The TensorFlow library was compiled to use AVX instructions, but these aren't available on your machine.
qemu: uncaught target signal 6 (Aborted) - core dumped
```

**Comment by sanschaise** [comment]
No update on this?

**Comment by josemiguelalves** [comment]
any update?

**Comment by harraz** [comment]
> I've tried `Tensorflow 2.3.1 ` and I still get `F tensorflow/core/lib/monitoring/sampler.cc:42] Check failed: bucket_limits_[i] > bucket_limits_[i - 1] (0 vs. 10) qemu: uncaught target signal 6 (Aborted) - core dumped` Any suggestions would be great - thanks.

Any luck with this issue. I get this when i try to import tensorflow in python

**Comment by dwyatte** [comment]
While this issue was originally opened around emulating TensorFlow on x86_64  in Docker, it does look like there are now `tensorflow` aarch64 binaries that can be used in linux/arm64/v8 Docker containers. More info here: https://blog.tensorflow.org/2022/09/announcing-tensorflow-official-build-collaborators.html

Dockerfile
```
FROM python:3.7-slim

RUN pip install tensorflow==2.10.0 tensorflow-io==0.27.0
CMD python -c "import tensorflow as tf; print(tf.constant(42) / 2 + 2)"
```

```
docker build --platform=linux/arm64/v8 . -t tensorflow
docker run --platform=linux/arm64/v8 tensorflow 
```

```
tf.Tensor(23.0, shape=(), dtype=float64)
```

**Comment by sachinprasadhs** [comment]
@dwyatte , Thanks for confirming, if your issue is resolved, could you please close this issue. 
Also, refer https://www.tensorflow.org/install for latest install instructions. Thanks!

**Comment by dwyatte** [comment]
> @dwyatte , Thanks for confirming, if your issue is resolved, could you please close this issue.

Sure, I think we can close this now. QEMU also appears to have merged AVX instructions, so once that is pulled into Docker, it might also be possible to run via emulation.

https://gitlab.com/qemu-project/qemu/-/issues/164#note_1140802183

**Comment by fumoboy007** [comment]
@sachinprasadhs Will Google release prebuilt ARM64 Docker images to [Docker Hub](https://hub.docker.com/r/tensorflow/tensorflow/)?

**Comment by sachinprasadhs** [comment]
CC:@angerson , @learning-to-play

**Comment by learning-to-play** [comment]
Thanks for reaching out! I'm not aware of any plans to release prebuilt ARM64 Docker images.

**Comment by fumoboy007** [comment]
@learning-to-play It would be great for the community if we had prebuilt images for all architectures that we support. 🙏

**Comment by Zhao-Dongyu** [comment]
I can’t believe this issue still exists in 2024. I tried installing TensorFlow with Anaconda, Docker, and specific TensorFlow versions, but none worked. Finally, this is how I solved it:

## Step 1: Install Miniforge

Miniforge is optimized for Apple Silicon.
```
curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh
bash Miniforge3-MacOSX-arm64.sh
source ~/miniforge3/bin/activate
```

## Step 2: Create a New Environment and Install TensorFlow

### 1.	Create and activate a new Conda environment:
```shell
conda create -n tensorflow_env python=3.10 -y
conda activate tensorflow_env
```

### 2.	Install TensorFlow:
```shell
conda install -c apple tensorflow-deps
python -m pip install tensorflow-macos
python -m pip install tensorflow-metal
```
## Step 3: Test the Installation

Run the following script to verify:
```shell
import tensorflow as tf
print("TensorFlow Version:", tf.__version__)
print("Is GPU available:", tf.config.list_physical_devices('GPU'))
```

**Comment by hkayann** [comment]
> I can’t believe this issue still exists in 2024. I tried installing TensorFlow with Anaconda, Docker, and specific TensorFlow versions, but none worked. Finally, this is how I solved it:
> 
> ## Step 1: Install Miniforge
> Miniforge is optimized for Apple Silicon.
> 
> ```
> curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh
> bash Miniforge3-MacOSX-arm64.sh
> source ~/miniforge3/bin/activate
> ```
> 
> ## Step 2: Create a New Environment and Install TensorFlow
> ### 1. Create and activate a new Conda environment:
> ```shell
> conda create -n tensorflow_env python=3.10 -y
> conda activate tensorflow_env
> ```
> 
> ### 2. Install TensorFlow:
> ```shell
> conda install -c apple tensorflow-deps
> python -m pip install tensorflow-macos
> python -m pip install tensorflow-metal
> ```
> 
> ## Step 3: Test the Installation
> Run the following script to verify:
> 
> ```shell
> import tensorflow as tf
> print("TensorFlow Version:", tf.__version__)
> print("Is GPU available:", tf.config.list_physical_devices('GPU'))
> ```

This issue is about dockers, please check the title.
